web-common/src/features/workspaces/CanvasWorkspace.svelte

```diff
@@ -0,0 +1,196 @@
+
+  $: canvasResource = data?.canvas;
+  $: canvasName = getNameFromFile(filePath);
+
+  $: metricsViewQuery = useDefaultMetrics(instanceId);
+
+  $: ({ instanceId } = $runtime);
+
+  $: lineBasedRuntimeErrors = mapParseErrorsToLines(
+    allErrors,
+    $remoteContent ?? "",
+  );
+
+  $: mainError = lineBasedRuntimeErrors?.at(0);
+
+  async function onChangeCallback(newTitle: string) {
+    const newRoute = await handleEntityRename(
+      $runtime.instanceId,
+      newTitle,
+      filePath,
+      fileName,
+    );
+    if (newRoute) await goto(newRoute);
+  }
+
+  async function addComponent(componentName: CanvasComponentType) {
+    const defaultMetrics = $metricsViewQuery?.data;
+    if (!defaultMetrics) return;
+
+    const newSpec = componentRegistry[componentName].newComponentSpec(
+      defaultMetrics.metricsView,
+      defaultMetrics.measure,
+      defaultMetrics.dimension,
+    );
+
+    const { width, height } = componentRegistry[componentName].defaultSize;
+    const newComponent = {
+      component: { [componentName]: newSpec },
+      height,
+      width,
+      x: 0,
+      y: 0,
+    };
+    const parsedDocument = parseDocument(
+      $editorContent ?? $remoteContent ?? "",
+    );
+
+    const items = parsedDocument.get("items") as any;
+
+    if (!items) {
+      parsedDocument.set("items", [newComponent]);
+    } else {
+      items.add(newComponent);
+    }
+
+    updateEditorContent(parsedDocument.toString(), true);
+    await updateComponentFile();
+  }
+</script>
+
+{#if canvasResource && fileArtifact}
+  {#key canvasName}
+    <StateManagersProvider {canvasName}>
+      <CanvasThemeProvider>
+        <WorkspaceContainer>
+          <WorkspaceHeader
+            slot="header"
+            {filePath}
+            hasUnsavedChanges={$hasUnsavedChanges}
+            titleInput={fileName}
+            onTitleChange={onChangeCallback}
+            resourceKind={ResourceKind.Canvas}
+          >
+            <div class="flex gap-x-2" slot="cta">
+              <PreviewButton
+                href="/canvas/{canvasName}"
+                disabled={allErrors.length > 0 || resourceIsReconciling}
+                reconciling={resourceIsReconciling}
+              />
+
+              <AddComponentMenu {addComponent} />
+              <ViewSelector
+                allowSplit={false}
+                bind:selectedView={$selectedViewStore}
+              />
+            </div>
+          </WorkspaceHeader>
+
+          <WorkspaceEditorContainer
+            slot="body"
+            error={mainError}
+            showError={!!$remoteContent && selectedView === "code"}
+          >
+            {#if selectedView === "code"}
+              <CanvasEditor
+                bind:autoSave={$autoSave}
+                {canvasName}
+                {fileArtifact}
+                {lineBasedRuntimeErrors}
+              />
+            {:else if selectedView === "viz"}
+              {#if mainError}
+                <ErrorPage
```

Comment:
Same here. We should show the error at the bottom of the page and still keep the editor up so that the user can fix issues.

Reply:
This seems to be fine right? The main error page is only shown when selected view is `viz` ?